Work on phase one of resolving issue 53. Add 2 instruction types to K…

Derek52 opened the pull request:
I think this is a workable implementation to get started with this issue.

There's also a syntax detail we will need to make a decision on. If you look at CacheJSTest.kt you will see
`browser.executeFromCache("""alert("Hello " + {});""","Derek")`

The syntax you put in the issue description, won't work. Not without some modifications at least.
`browser.executeFromCache("""alert("Hello {}");""","Derek")`

That doesn't work, because right now, I'm just swapping out appearance of "{}" for randomly generated variable names. But, with the syntax written that way, inside the quote marks, javascript won't recognize the variable name. 

So, Kweb users will need to treat "{}" as variable names and format their strings correctly, i.e., `alert("Hello" +  {})`, or we will need to add some complexity and "break out" variables. We could parse "Hello {}, how are you?", and change it into "Hello " + variableName + ", how are you?" But, that's more complex then the basic code I wrote to substitute the "{}" so far.

sanity commented:
> So, Kweb users will need to treat "{}" as variable names and format their strings correctly

I think this is a reasonable requirement, we should just be sure to document it in the kdoc for executeFromCache().

Derek52 commented:
I will be sure to document it. This functionality should be moved to execute() though. executeFromCache() is temporary.

As for the next step here, I'm going to work on the instruction infrastructure next. Right now we have the data class `Server2ClientMessage`. The class fields are `yourId`, `debugToken`, `execute`, `evaluate`, and `instructions`.

What do you think this class should look like after the rewrite? Should I make instructions a data class that stores the javascript and arguments? Should I remove instructions entirely and try to cache/execute JS using the execute field of the Message?

sanity commented:
> What do you think this class should look like after the rewrite? 

I think we can simplify it, `yourId` is probably still needed as is `debugToken`, it might be good if we could consolodate `execute` and `evaluate`, since `execute` is essentially a special-case of `evaluate` where no value is returned.  Per previous discussion, this mechanism should replace `instructions` with something more general.

> Should I make instructions a data class that stores the javascript and arguments? 

I think these changes should make `instructions` obsolete.

> Should I remove instructions entirely and try to cache/execute JS using the execute field of the Message?

Right, I think the `Server2ClientMessage` could look something like:

```kotlin
data class Server2ClientMessage(
  val yourId : String, // Same as before
  val debugToken : String?,  // Same as before
  val jsId : Int, // The id of the javascript snippet, chosen by server
  val js : String?, // Null if this should have already been cached under the jsId
  val parameters : List<Any>,
)
```

Derek52 commented:
Ok, cool. That's about what I had in mind. Except I had the idea to put js and parameters into a data class called Instruction. I don't know why I had that in my head. I like it more the way you have it.

Derek52 commented:
https://github.com/Derek52/kweb-core/blob/73d9a75e94b3bf031025116a7a4bbd3544a3401f/src/main/kotlin/kweb/WebBrowser.kt#L118-L120

I ran into an issue with this function today. I'm a little confused by what `rh` is, and what it's doing.

Also, I've run into an issue with the callback system in general. Currently, the `callbackId` is being sent in the Evaluate data class. I got rid of that class. I think I'd like to make `callbackId` a part of the `Server2ClientMessage` and give it a default value of `-1`. Then in js on the client side, I can check the id, ignore it if it is `-1`, or send the result back to the server if the id isn't `-1`.

Does that seem like a good solution to you? And is `-1` a safe value for that? I haven't checked yet to see if a `callbackId` could be set to `-1` or not yet.

sanity commented:
> I ran into an issue with this function today. I'm a little confused by what rh is, and what it's doing.

`rh` is a callback, I think it's probably short for "result handler" (which would be a better name).  When the response is received from the client it is passed to the result handler.

> Does that seem like a good solution to you? And is -1 a safe value for that? I haven't checked yet to see if a callbackId could be set to -1 or not yet.

Hi, I think `-1` could work if we verify that those ids can't be negative (they are probably just randomly generated) although `null` might be preferable.

Derek52 commented:
I like to avoid making fields nullable, especially primitives. I think it just looks nicer having fields be non-nullable. After thinking about it, I think making the callbackId nullable probably is the best solution here though.

Derek52 commented:
I refactored my javascript and now all but the 3 Todo demo tests pass. And I've run into my first real headache here.

If you look at the change I made to `ElementCreator` in the last commit, you'll see I had to add the word "return" to the JS expression. This is because when calling `func.apply()` in the JS, it wasn't returning any data. Because the js wasn't set to return anything. This was causing my StringDiff tests to fail, because they couldn't get the value of the text in an input field.

So far, I don't know if adding "return" in that one place is a good enough fix. I don't know if that can break anything anywhere else, and I think it feels like a hack.

Also, the TodoDemoTests are failing, for what I think is a similar reason. That test file has a method called `getItemByText` and based on the errors Selenium is spitting out, I think that method also relies on a js expression somewhere without a return statement.

I'm interested on if you know how to proceed from here.

Also, do we want to cache ALL javascript? Because I've run into a couple of cases that are gonna make that tricky. The first is Kweb's `refreshAllPages` method. https://github.com/Derek52/kweb-core/blob/6d45809f36efc183a5839527d7d69277081e7b85/src/main/kotlin/kweb/Kweb.kt#L430-L439

I've added a special case to `handleInboundMessage` for cases like `refreshAllPages`. If the `cacheId` is null, it will just execute the javascript and not worry about the cache. In the case of `refreshAllPages`, to be able to cache that function, I think the best way would be to hardcode that function in the javascript, and manually assign it a cacheId in the js, and in the kotlin function where it's called.

sanity commented:
You want to jump on a video chat tomorrow and we can work through it?  Grab a time here: https://cal.mixmax.com/sanity/55

Derek52 commented:
A bunch of commits have been pushed. I was gonna link to a couple of points I'd like to highlight, but I decided to just leave TODO's and comments on them instead.

sanity commented:
I've made a few comments but overall this looks great. Once we're happy with it the next step is probably to implement function pre-caching, so that functions called during page render are substituted into kweb_bootstrap.js, similar to the `<!-- BUILD PAGE PAYLOAD PLACEHOLDER -->`, but where they are parsed as part of kweb_bootstrap.js rather than by using `new Function()`, this should be more efficient from a browser CPU perspective.

Derek52 commented:
Where are your comments? I'm not seeing anything.

And I'm not sure what functions you are talking about pre-caching. Can you link to one, or a section of code with a few of them?

sanity commented:
Hopefully you can see them here: https://github.com/kwebio/kweb-core/pull/180/files

Derek52 commented:
I'm not seeing any comments. Do you need to push a commit?

Derek52 commented on src/main/kotlin/kweb/html/ElementReader.kt:
```diff
@@ -19,15 +19,15 @@ open class ElementReader(protected val receiver: WebBrowser, internal val jsExpr
-    val tagName: CompletableFuture<String> get() = receiver.evaluate("$jsExpression.tagName").thenApply { it.toString() }
-    val attributes: CompletableFuture<Map<String, Any>> get() = receiver.evaluate("$jsExpression.attributes").thenApply { it as Map<String, Any> }
-    fun attribute(name: String): CompletableFuture<Any> = receiver.evaluate("($jsExpression.getAttribute(\"${name.escapeEcma()}\"));")
+    val tagName: CompletableFuture<String> get() = receiver.evaluate("return $jsExpression.tagName").thenApply { it.toString() }
```
Right now, the client has no means of evaluating anything. It can only call functions. Without these return statements, those expressions won't be able to return data. This is the trade-off caused by removing eval() calls from the javascript. Using javascript's eval(), would let us evaluate expressions. We'd have to add another field to the Server2ClientMessage class though, so the client would know if it was receiving an expression to call eval() on, or a function to call.

sanity replied:
Ah, yes, that makes sense, it's the body of a function. We might want to rename `callJs` to make this more explicit, like `callJsFunction` calling the parameter `body`.

Derek52 commented on src/main/kotlin/kweb/WebBrowser.kt:
```diff
@@ -59,37 +65,79 @@ class WebBrowser(private val sessionId: String, val httpRequestInfo: HttpRequest
+            params.append("$jsVarName,")
+        }
+        if (params.isNotEmpty()) {
+            params.deleteCharAt(params.lastIndex)//delete the last trailing comma
```
Yes we could.

Derek52 commented on src/main/kotlin/kweb/Kweb.kt:
```diff
@@ -152,15 +152,38 @@ class Kweb private constructor(
         }
     }
 
+    fun cacheAndExecute(clientId: String, cacheId: Int, js: String, parameters: String, vararg args: Any) {
```
You were looking at the wrong commit here. Both of those functions are gone. I sorted this problem out already.

Derek52 commented on src/main/kotlin/kweb/client/Server2ClientMessage.kt:
```diff
@@ -2,22 +2,13 @@ package kweb.client
-        val instructions: List<Instruction>? = null
-) {
+        var debugToken: String? = null,
+        val jsId: Int? = null, //the id used to store or receive a js function in cache.
```
I agree. That's one of the last things I like to do though. I like things to be a little more finalized before I take the time to do that.

Derek52 commented on src/main/kotlin/kweb/WebBrowser.kt:
```diff
@@ -59,37 +65,79 @@ class WebBrowser(private val sessionId: String, val httpRequestInfo: HttpRequest
+    I'm not 100% sure what to do with this function. I think it'd be too complicated to merge it into
+    one of the callJs functions.
+    Maybe it should just be named something like 'callJsWithResult' or something.*/
+    fun evaluate(js: String, vararg args: Any?): CompletableFuture<Any> {
```
I remember talking about this. I forgot you mentioned that. I just tried rewriting it as a suspend function. I then went to the first place it's called in CookieReceiver, and was getting an error. Also, that call is written to handle a CompletableFuture.
This will probably have to work as a suspend function. But, I'm gonna have to play with it for a little while, and probably rewrite some code in the places where it is currently called.

Derek52 commented on src/main/kotlin/kweb/html/fileUpload/FileFormInput.kt:
```diff
@@ -21,8 +21,8 @@ class FileFormInput {
 
-    fun setAccept(acceptedTypes: String): Unit = inputElement.execute("${inputElement.jsExpression}.accept=\"$acceptedTypes\";")
-    fun isMultiSelect(isMultiple: Boolean): Unit = inputElement.execute("${inputElement.jsExpression}.multiple=\"$isMultiple\";")
+    fun setAccept(acceptedTypes: String): Unit = inputElement.callJs("${inputElement.jsExpression}.accept=\"$acceptedTypes\";")
```
Something like `callJs("document.getElementById({}).accept = {}", inputElement.id, acceptedTypes)`?
Because that should work. I was just so focused on changing all the instances of execute to callJs, that I overlooked adding our new variable syntax. I'll go back through and fix these this evening.

Derek52 commented on build.gradle:
```diff
@@ -25,7 +25,7 @@ ext {
 
 group 'com.github.kwebio'
-version '0.8.5'
+version '0.8.6'
```
I bumped it to 0.8.6 to match the current release number. I agree it should be changed to 0.9.0. But, that's another thing I don't really worry about until things are more final.

sanity commented on src/main/kotlin/kweb/WebBrowser.kt:
```diff
@@ -59,37 +65,79 @@ class WebBrowser(private val sessionId: String, val httpRequestInfo: HttpRequest
+        val params = StringBuilder()
+        while (js.contains("{}")) {
+            //a few random letters, an underscore, and a randomly generated number should make a variable name that no one would ever come up with
+            val jsVarName = "rzd_${rng.nextInt(1000)}"
```
Might want to increase to nextInt(100000) to reduce chance of collisions.

sanity replied:
Alternatively we could use code similar to WebBrowser.generateId() (but we probably don't want to use generateId() itself because it's useful for element ids to be sequential)

sanity commented on src/main/kotlin/kweb/WebBrowser.kt:
```diff
@@ -163,7 +156,7 @@ class WebBrowser(private val sessionId: String, val httpRequestInfo: HttpRequest
         }
-        callJs("""
+        callJsFunction("""
         history.pushState({}, "", "$url");
```
Should $url be a parameter here?

sanity commented on src/main/kotlin/kweb/Element.kt:
```diff
@@ -299,42 +305,44 @@ open class Element(
+                    let value = {}
+                    var ntn=document.createTextNode(value);
+                    document.getElementById(id).appendChild(ntn);
+                """, id, value.escapeEcma())
```
Might want to use https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/trim-indent.html here.

sanity commented on src/main/kotlin/kweb/WebBrowser.kt:
```diff
@@ -70,14 +70,23 @@ class WebBrowser(private val sessionId: String, val httpRequestInfo: HttpRequest
-            val jsVarName = "auto_var_${params.size+1}"
-            js = js.replaceFirst("{}", jsVarName)
-            params.add(jsVarName)
+        var i = 0
```
Looks good.

Derek52 commented:
If you look at my last commit, you'll see my problem with text being auto escaped (i.e., "We'll" became "We\'ll"), was fixed by removing .escapeEcma() from the value in Element.text()

I did a little googling, and I'm not sure what escapeEcma() does or why you were using it. I know that removing .escapeEcma() makes the todo demo app work for inputs with things like single quote marks '.

I'm not sure what types of strings that would no longer work with escapeEcma() not being used. This is probably a design decision I'm not quite informed enough to make myself.

Derek52 commented:
DIsregard that above comment. I realized that the issue with escapeEcma() is that you were using it to insert a value directly into some JavaScript code. When sending the value as a function argument, the escapeEcma isn't necessary.

sanity commented:
Fixes #53
